From 528d1dcdd1785c5797a19536ab348467844256b7 Mon Sep 17 00:00:00 2001
From: Ben Wagner <bungeman@google.com>
Date: Mon, 12 Aug 2024 15:00:08 -0400
Subject: [PATCH] [pdf] Bounds check in skia_alloc_func

The allocator callback for zlib needs to check that items * size will
fit in size_t and return nullptr if not.

Conflicts:
- src/pdf/SkDeflate.cpp: just in header includes

Bug: 349678452
Reviewed-on: https://skia-review.googlesource.com/c/skia/+/888996
Commit-Queue: Ben Wagner <bungeman@google.com>
Reviewed-by: Brian Osman <brianosman@google.com>
(cherry picked from https://googleplex-android-review.googlesource.com/q/commit:c69cda562fe7936c65e0844acc09e927d73ccfce)
Merged-In: Id1a30592d435bd0de4630e7047f26b0dc17654fc
Change-Id: Id1a30592d435bd0de4630e7047f26b0dc17654fc
---
 src/pdf/SkDeflate.cpp | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/src/pdf/SkDeflate.cpp b/src/pdf/SkDeflate.cpp
index 2c08f12aab..265f3b4e02 100644
--- a/src/pdf/SkDeflate.cpp
+++ b/src/pdf/SkDeflate.cpp
@@ -10,6 +10,7 @@
 #include "include/private/base/SkAssert.h"
 #include "include/private/base/SkDebug.h"
 #include "include/private/base/SkMalloc.h"
+#include "include/private/base/SkTFitsIn.h"
 #include "include/private/base/SkTo.h"
 #include "src/core/SkTraceEvent.h"
 
@@ -24,6 +25,13 @@ namespace {
 // Different zlib implementations use different T.
 // We've seen size_t and unsigned.
 template <typename T> void* skia_alloc_func(void*, T items, T size) {
+    if (!SkTFitsIn<size_t>(size)) {
+        return nullptr;
+    }
+    const size_t maxItems = SIZE_MAX / size;
+    if (maxItems < items) {
+        return nullptr;
+    }
     return sk_calloc_throw(SkToSizeT(items) * SkToSizeT(size));
 }
 
-- 
2.46.1.824.gd892dcdcdd-goog

